### ISSUE TITLE
alttext+speechtext leads to duplication [New feature request : Make alttext optional ]

### ISSUE THREAD

**Issue by nirupa-kumar** [mathjax/MathJax-node#121]
New feature request : Make alttext optional when providing mml output for equation conversion. 
Currently when we have the mml output provided , this gives us the alttext as default. This leads to equations being read out twice with AT devices. This does not allow browsers and readers to make use of the a11y features for MathML.  It would be nice to have it as an optional component. 
As an additional feature , if Latex code were an option for alttext that would be great.

**Comment by dpvc** [comment]
It is not clear which command or API you are using, but the various command-line tools in the `bin` directory only include the speech text as `alttext` if the `--speech` option is included on the command (or if there is already an `alttext` attribute on the input MathML).  If you are calling the API from your own node program, you can use the `speakText` property being passed to the MathJax `typeset()` command to control whether the `alttext` is included or not.  So you can already control whether speech text is added or not both in the command-line functions and the direct API calls.

It's true that there is not an option for including the TeX code as `alttext`, but if the `alttext` is being read by screen readers, doesn't that mean that both the TeX and the MathML will be read?  That seems to be just as bad.  And for anything but the most simple TeX expressions, reading the TeX code would produce pretty poor results for people wanting to understand the math.  That's one of the reasons we don't do that.  If you want the TeX to be available for other machine-reading purposes, however, you can use the `--semantics` command-line option to have the TeX included in a `<semantics>` annotation to the MathML where it can be processed by other tools that read MathML.

Finally, shouldn't the `alttext` only be read if the AT doesn't have a means of reading the actual content of the tag (so if it understands how to read MathML, it should *not* read the `alttext`)?

**Comment by nirupa-kumar** [comment]
Oddly enough , it seems to be reading both. We are using the APIs for our node.js application. I maybe completely wrong here. But I was going over the mj-single.js and mj-page.js codes and this is what I came across : 
In mj-page.js : 
```javascript
function AdjustMML() {
  if (data.renderer === "NativeMML") {
    var nodes = document.getElementsByClassName("MathJax_MathML");
    for (var i = nodes.length-1; i >= 0; i--) {
      var math = nodes[i].getElementsByTagName("math")[0];
      var jax = MathJax.Hub.getJaxFor(math);
      math.setAttribute("id", jax.inputID);
      nodes[i].parentNode.replaceChild(math,nodes[i]);
      var description = speech.processExpression(math.outerHTML);
      jax.speech = description;
      if (data.speakText) math.setAttribute("alttext",description);
    }
  }
}
```
Here , from what I understand , it refers to the speakText for all conversions (i.e. mml , svg and png ). 

And in case of mj-single.js : 
```javascript
function GetSpeech(result) {
  if (!data.speakText) return;
  speech.setupEngine({semantics: true, domain: data.speakRuleset, style: data.speakStyle});
  result.speakText = speech.processExpression(result.mml);
  if (data.mml) {
    var jax = MathJax.Hub.getAllJax()[0];
    jax.root.alttext = result.speakText;
    if (jax.root.attrNames) {jax.root.attrNames.push("alttext")}
    result.mml = jax.root.toMathML('',jax);
  }
}
```
Similarly in the above too. Please let me know if my understanding is correct. That's what led me to believe that the alttext may not be optional. We generate all the formats for each equation submitted. So we have the text , svg , png and mml outputs produced.

**Comment by pkra** [comment]
Thanks for the follow up, @nirupa-kumar.

> That's what led me to believe that the alttext may not be optional. 

Yes, it looks like we're forcing SRE's output as alt-text on MathML output. We probably shouldn't. @dpvc?

**Comment by dpvc** [comment]
> it looks like we're forcing SRE's output as alt-text on MathML output

This is only when speech text is requested.  In that case, the speech is added to the output formats like MathML and SVG (by way of the MathML).  That is so that something like `bin/tex2mml --speech` will end up with the speech text in the result.

If you want to consider the speech text to be a separate output format, then either the driver files would be responsible for adding the speech to the other formats, or there needs to be a separate API flag controlling whether the speech is to be combined with the output or not.

I don't like the first option very well, as that leads to duplicate (and possibly inconsistent) code in all the drivers.  Also, since the MathML and SVG are returned as text strings, it means reprising, inserting the speech into the node, and the reserializing the result.  I would like to reduce the parsing/serializing as much as possible, as that has been a performance issue in the past.

**Comment by pkra** [comment]
> If you want to consider the speech text to be a separate output format, then either the driver files would be responsible for adding the speech to the other formats, or there needs to be a separate API flag controlling whether the speech is to be combined with the output or not.

Personally, I think speech-generation should meet the fate of PNG-generation, i.e., be separated out.

**Comment by dpvc** [comment]
> I think speech-generation should meet the fate of PNG-generation, i.e., be separated out.

Can you give an idea of the work-flow you have in mind?

Also, the speech addition was one of the original design requirements for mathjax-node (from the project that originally funded it), which is why it was there.  I'm OK with changing that now, but didn't want to forget the history.

**Comment by pkra** [comment]
> Can you give an idea of the work-flow you have in mind?

Hm. Just to jot down something, I could imagine writing something like

```javascript
var mj = require('mj-single');
var sre = require('speech-rule-engine');

exports.mjSre = function(options, callback) {
  sre.setupEngine(options.sre);
  mj.start();
  mj.typeset(options.mj, function(data) {
    data.speakText = sre.processExpression(data.mml);
    if (options.svg){
      // modify data.svg
    }
    // etc
    callback(data);
  });
};
```


> Also, the speech addition was one of the original design requirements for mathjax-node (from the project that originally funded it), which is why it was there. I'm OK with changing that now, but didn't want to forget the history.

Of course. I should be clearer about that kind of thing -- sorry. The original requirements were what they were but I think (especially with MathJax v3.0 in mind) pulling things apart and seeing what we (and others) actually need in mathjax-node itself would be useful. And micro-modules have benefits of their own.

**Comment by dpvc** [comment]
The line I'm concerned about is 

    // modify data.svg

since `data.svg` is a string, not a DOM element, so modifying it requires re-parsing and re-serializing, which is less efficient than adding it when it is already a DOM element.

It might be possible to return a DOM element instead (I haven't looked into whether it is OK to pass parts of the jsdom DOM back to the caller or not, since the DOM used in mj-single is private).

**Comment by pkra** [comment]
Oh and

> This is only when speech text is requested.

Yes. Sorry about missing that :frowning:

**Comment by pkra** [comment]
> The line I'm concerned about is
> `// modify data.svg`
> since data.svg is a string, not a DOM element, so modifying it requires re-parsing and re-serializing, > which is less efficient than adding it when it is already a DOM element. 

I agree. Something to be worked out.

As for the OP, I've noticed something strange


```html
$ node bin/mml2mml --speech '<math alttext="hello world"> <msup> <mi>&#x03B1;<!-- α --></mi> <mn>2</mn> </msup> </math>'
<math xmlns="http://www.w3.org/1998/Math/MathML" alttext="alpha squared" alttext="alpha squared">
  <msup>
    <mi>&#x3B1;</mi>
    <mn>2</mn>
  </msup>
</math>
```

and 

```html
$ node bin/mml2svg --speech '<math alttext="hello world"> <msup> <mi>&#x03B1;<!-- α --></mi> <mn>2</mn> </msup> </math>'
<svg xmlns:xlink="http://www.w3.org/1999/xlink" width="2.542ex" height="2.509ex" style="vertical-align: -0.338ex;" viewBox="0 -934.9 1094.4 1080.4" aria-label="hello world" role="math" focusable="false" xmlns="http://www.w3.org/2000/svg" aria-labelledby="MathJax-SVG-1-Title MathJax-SVG-1-Desc">
<title id="MathJax-SVG-1-Title">Equation</title>
<desc id="MathJax-SVG-1-Desc">alpha squared</desc>
<defs aria-hidden="true">
<path stroke-width="1" id="E1-MJMATHI-3B1" d="M34 156Q34 270 120 356T309 442Q379 442 421 402T478 304Q484 275 485 237V208Q534 282 560 374Q564 388 566 390T582 393Q603 393 603 385Q603 376 594 346T558 261T497 161L486 147L487 123Q489 67 495 47T514 26Q528 28 540 37T557 60Q559 67 562 68T577 70Q597 70 597 62Q597 56 591 43Q579 19 556 5T512 -10H505Q438 -10 414 62L411 69L400 61Q390 53 370 41T325 18T267 -2T203 -11Q124 -11 79 39T34 156ZM208 26Q257 26 306 47T379 90L403 112Q401 255 396 290Q382 405 304 405Q235 405 183 332Q156 292 139 224T121 120Q121 71 146 49T208 26Z"></path>
<path stroke-width="1" id="E1-MJMAIN-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path>
</defs>
<g stroke="currentColor" fill="currentColor" stroke-width="0" transform="matrix(1 0 0 -1 0 0)" aria-hidden="true">
 <use xlink:href="#E1-MJMATHI-3B1" x="0" y="0"></use>
 <use transform="scale(0.707)" xlink:href="#E1-MJMAIN-32" x="905" y="513"></use>
</g>
</svg>
```

Not sure if that's what the OP had in mind but this strikes me as not intended.

**Comment by dpvc** [comment]
OK, that is a good point:  `alttext` appearing as `aria-label` but speech as `aria-labeled-by`.  Not sure if that is what the OP had in mind, but it probably should not be both.

**Comment by pkra** [comment]
Right. Note also the duplicate alttext in the MathML output (caused by a missing check in https://github.com/mathjax/MathJax-node/blob/master/lib/mj-single.js#L544, I think, but that doesn't fix the SVG issue so I'm guessing the original alttext is stored somewhere else as well?)

**Comment by dpvc** [comment]
Thanks, I missed the duplication in the MathML.  You are right about the line that causes that.

**Comment by pkra** [comment]
Renamed so that we can address the bugs that turned up in the discussion.

**Comment by pkra** [comment]
Closing this in favor of #236
